### DISCUSSION TITLE
[BUG]: Duplicate joins are being generated in SQL queries when populating relationships using

### DISCUSSION THREAD

**Discussion by tufailra97** [mikro-orm/mikro-orm#5934] [General]
When attempting to populate related entities, the generated SQL query includes multiple duplicate JOIN clauses. This results in unnecessarily complex and inefficient SQL, as the same tables are joined multiple times under different aliases. The issue occurs particularly when populating relationships and applying filters on related entities.

**Example:**

Given the following entities:
```ts
@Entity({ tableName: 'users' })
export class UserEntity extends BaseEntity {
  @Property({ name: 'first_name' })
  firstName!: string;

  @Property({ name: 'last_name' })
  lastName!: string;

  @Formula(`CONCAT(first_name, ' ', last_name)`)
  fullName!: string;

  @Index()
  @OneToMany(() => ProfileEntity, (userProfile) => userProfile.user)
  userProfiles = new Collection<ProfileEntity>(this);
}

@Entity({ tableName: 'roles' })
export class RoleEntity extends BaseEntity {
  @Property()
  name!: string;

  @Index()
  @OneToMany(() => ProfileEntity, (user) => user.role)
  users = new Collection<ProfileEntity>(this);
}

@Entity({ tableName: 'profiles' })
export class ProfileEntity extends BaseEntity {
  @Property({ nullable: true })
  username!: string;

  @Property({ nullable: true })
  imageUrl!: string;

  @Property({ default: true, name: 'is_active' })
  isActive!: boolean;

  @ManyToOne()
  user!: Rel<UserEntity>;

  @ManyToOne()
  role!: Rel<RoleEntity>;
}
```

When querying with a populated relationship and additional filters:

```ts
const [users, total] = await orm.em.findAndCount(
    UserEntity,
    {
      fullName: {
        $ilike: '%john%',
      },
      userProfiles: {
        isActive: true,
        role: {
          name: {
            $ilike: '%admin%',
          },
        },
      },
    },
    {
      populate: ['userProfiles.role'],
    }
  );
```

The ORM generates the following SQL queyr:
```sql
select
  "u0".*,
  "u1"."id" as "u1__id",
  "u1"."created_at" as "u1__created_at",
  "u1"."updated_at" as "u1__updated_at",
  "u1"."deleted_at" as "u1__deleted_at",
  "u1"."username" as "u1__username",
  "u1"."image_url" as "u1__image_url",
  "u1"."is_active" as "u1__is_active",
  "u1"."user_id" as "u1__user_id",
  "u1"."role_id" as "u1__role_id",
  "r2"."id" as "r2__id",
  "r2"."created_at" as "r2__created_at",
  "r2"."updated_at" as "r2__updated_at",
  "r2"."deleted_at" as "r2__deleted_at",
  "r2"."name" as "r2__name",
  CONCAT(first_name, ' ', last_name) as "full_name"
from
  "users" as "u0"
  left join "profiles" as "u1" on "u0"."id" = "u1"."user_id"
  left join "roles" as "r2" on "u1"."role_id" = "r2"."id"
  left join "profiles" as "p3" on "u0"."id" = "p3"."user_id"
  left join "roles" as "r4" on "p3"."role_id" = "r4"."id"
where
  CONCAT(first_name, ' ', last_name) ilike '%john%'
  and "p3"."is_active" = true
  and "r4"."name" ilike '%admin%'
```
As seen above, the query includes duplicate joins for the profiles and roles tables, resulting in redundant SQL and potentially degraded performance.

**Notes:**
I am currently using `v6.3.4` but I noticed with older version too.

### Reproduction

https://github.com/tufailra97/mikro-orm-duplicate-joins-repro

### What driver are you using?

None

### MikroORM version

6.3.4

### Node.js version

20

### Operating system

Mac OS

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Discord](https://discord.gg/w8bjxFHS7X).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
This is expected behavior, they are not duplicated, one branch is used for population (loading the values), and the other is for where condition. You can see one alias is used in the select clause, while the other is used only in the where clause. You can control this behavior with `populateWhere` option, it defaults to `all` which means separate joins are used for loading values since the where query is not supposed to affect them. Use `infer` if you want to reuse the existing join branches, as long as you actually want this - if you want to filter the relations based on your query too. Alternatively, you can put a custom condition to the `populateWhere` option.

https://mikro-orm.io/docs/loading-strategies#population-where-condition

Maybe you want something like this:

```ts
const [users, total] = await orm.em.findAndCount(
    UserEntity,
    {
      fullName: {
        $ilike: '%john%',
      },
    },
    {
      populate: ['userProfiles.role'],
      populateWhere: {
        userProfiles: {
          isActive: true,
          role: {
            name: {
              $ilike: '%admin%',
            },
          },
        },
      },
    },
  );
```

**Reply by tufailra97** [reply]
Excuse my ignorance, but don't duplicate joins lead to performance issues?

**Reply by B4nan** [reply]
They are not duplicate, one is for select, one is for where, not having them like this would produce different values.

**Reply by tufailra97** [reply]
I see, make sense. Thanks!

**Reply by artsiommiksiuk** [reply]
@B4nan, I'm not sure I'm following here.

If you doing left join, you have access to the whole row after that. Independently of what is written in a select clause, you may use everything from joined row for filtering. I don't quite understand this whole notion of join for select and join for where clause. It's the first time I see such logic. Like, the intent is clear, while reasoning is not at all, because it doesn't make sense. I'm not sure what you mean by different values either in this case.

I'm speaking mostly in the context of SQL databases (postgresql), so I may not know something. E.g. I know this is an issue for GCP Datastore, where everything you filtering by needs to be in select clause, but this is very specific and limited cloud database.

Can you elaborate a bit more or point me if it were described somewhere a bit more elaborated?

**Reply by B4nan** [reply]
This has been discussed several times already, the defaults are `populateWhere: 'all'`, which means your where query will not affect the population. In other words, if you do `em.find(User, { books: { title: 'foo' } }, { populate: ['books'] })`, you will end up with users that have a book called `foo`, but their `books` collection will contain all their books. This is why there need to be two join branches for the books relation, since if we had just one, you would end up with only one book in the collection (called `foo`).
